pkg/apis/networking/types.go

```diff
@@ -246,16 +246,39 @@ type IngressList struct {
 
 // IngressSpec describes the Ingress the user wishes to exist.
 type IngressSpec struct {
-	// IngressClassName is the name of the IngressClass cluster resource. The
-	// associated IngressClass defines which controller will implement the
-	// resource. This replaces the deprecated `kubernetes.io/ingress.class`
-	// annotation. For backwards compatibility, when that annotation is set, it
-	// must be given precedence over this field. The controller may emit a
-	// warning if the field and annotation have different values.
-	// Implementations of this API should ignore Ingresses without a class
-	// specified. An IngressClass resource may be marked as default, which can
-	// be used to set a default value for this field. For more information,
-	// refer to the IngressClass documentation.
+	// IngressClassName is the name of an IngressClass cluster resource.
+	// Ingress controller implementations use this field to know whether they
+	// should be picking up this Ingress.
+	//
+	// The field `ingressClassName` (referred to as “the field”) should be
+	// preferred over the annotation `kubernetes.io/ingress.class` (referred to
+	// as “the annotation”). The annotation and the field are different: the
+	// field refers to an IngressClass that exists in the cluster (although
+	// implementations may not require the presence of an IngressClass with a
+	// matching name). On the other side, the annotation is an opaque string
+	// that does not relate to the IngressClass type.
+	//
+	// The annotation and the field cannot be set simultaneously when an Ingress
+	// is created. In order to ease the process of progressively migrating from
+	// the annotation to the field, it is still possible to set both the
+	// annotation and field by updating an existing Ingress. This is useful for
+	// people migrating away from an implementation that does not support the
+	// field. Because of this possibility of having both the annotation and
+	// field, implementations must give the annotation precedence over the field
+	// when both are set. Kubernetes does not validate that both the annotation
+	// and field match when an Ingress is updated; having mismatched values may
+	// be useful for migrating. Implementations may warn the user of such a
+	// mismatch.
```

Comment:
I think it's very important to communicate that these values do not refer to the same thing. This text seems to suggest that they will largely be the same value, but I want to be careful to distinguish between the two. The annotation refers to a constant name known or specified in advanced by the controller. The field refers to the name of a GatewayClass resource, and that GatewayClass resource is tied to the controller with the included controllerName field. For example, I can see it being at least somewhat common to use generic GatewayClass names across different environments to provide a consistent UX, even if the underlying vendors are different. That would not be possible with the annotation.